### PR TITLE
[Spark] Support 3-part naming in table identifier parsing

### PR DESCRIPTION
**Pull request by gengliangwang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
Support 3-part naming in table identifier parsing. Before the changes, the following command 
```
OPTIMIZE catalog_foo.db.tbl
```
will throw error
```
org.apache.spark.sql.delta.DeltaParseException:
Illegal table name catalog_foo.db.tbl(line 1, pos 9)

== SQL ==
optimize catalog_foo.db.tbl
---------^^^

  at io.delta.sql.parser.DeltaSqlAstBuilder.$anonfun$visitTableIdentifier$1(DeltaSqlParser.scala:430)
  at org.apache.spark.sql.catalyst.parser.ParserUtils$.withOrigin(ParserUtils.scala:160)
  at io.delta.sql.parser.DeltaSqlAstBuilder.visitTableIdentifier(DeltaSqlParser.scala:427)
  at io.delta.sql.parser.DeltaSqlAstBuilder.$anonfun$visitOptimizeTable$5(DeltaSqlParser.scala:348)
  at scala.Option.map(Option.scala:230)
  at io.delta.sql.parser.DeltaSqlAstBuilder.$anonfun$visitOptimizeTable$1(DeltaSqlParser.scala:348)
```
After the changes, the command works.

## How was this patch tested?

A new unit test

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @tdas

**Comment by felipepessoto** [comment]
Is this specific for OPTIMIZE? Context: https://github.com/delta-io/delta/issues/1705#issuecomment-1581540251

**Review by tdas — APPROVED** [review]
LGTM. Thanks for catching and fixing this.

**Comment by vkorukanti** [comment]
> Is this specific for OPTIMIZE? Context: [#1705 (comment)](https://github.com/delta-io/delta/issues/1705#issuecomment-1581540251)

Not sure if I understand the qn correctly. This is applicable for all commands defined in the Delta SQL override (basically the commands using the `visitTableIdentifier`)

**Comment by vkorukanti** [comment]
@gengliangwang Could you please fix the test failures in suite `io.delta.sql.parser.DeltaSqlParserSuite`?

**Comment by gengliangwang** [comment]
> Is this specific for OPTIMIZE? Context: https://github.com/delta-io/delta/issues/1705#issuecomment-1581540251

It's for delta parser.  And with the changes of this one and https://github.com/delta-io/delta/issues/1705, optimize command now supports 3L name.

**Comment by gengliangwang** [comment]
> @gengliangwang Could you please fix the test failures in suite io.delta.sql.parser.DeltaSqlParserSuite?

@vkorukanti Yes I think it should be fixed now.

File: spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala [hunk 1]
```diff
@@ -260,11 +266,8 @@ class DeltaSqlParserSuite extends SparkFunSuite with SQLHelper {
     checkCloneStmt(parser, source = "t1", target = "t1", versionAsOf = Some(1L))
     // Clone with 3L table (only useful for Iceberg table now)
     checkCloneStmt(parser, source = "local.iceberg.table", target = "t1", sourceIs3LTable = true)
-    // Yet target cannot be a 3L table yet
-    intercept[ParseException] {
-      checkCloneStmt(parser, source = "local.iceberg.table", target = "catalog.delta.table",
-        sourceIs3LTable = true)
-    }
+    checkCloneStmt(parser, source = "local.iceberg.table", target = "delta.table",
```
Comment (gengliangwang) [inline spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala:269]: In https://github.com/delta-io/delta/blob/master/spark/src/main/scala/io/delta/sql/parser/DeltaSqlParser.scala#L287, the catalog is ignored. We need to fix it later.

Comment (vkorukanti) [reply spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala:269]: Could you please create a github issue?

Comment (gengliangwang) [reply spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala:269]: Sure https://github.com/delta-io/delta/issues/1992

**Comment by felipepessoto** [comment]
> > Is this specific for OPTIMIZE? Context: [#1705 (comment)](https://github.com/delta-io/delta/issues/1705#issuecomment-1581540251)
> 
> Not sure if I understand the qn correctly. This is applicable for all commands defined in the Delta SQL override (basically the commands using the `visitTableIdentifier`)

It seems we have more than one implementation of table resolution. The question is if the problem exists for other commands (DESCRIBE HISTORY for example), and if the fix would fix them all.

**Comment by gengliangwang** [comment]
> The question is if the problem exists for other commands (DESCRIBE HISTORY for example), and if the fix would fix them all.

1. The problem exists in other commands. I don't have a full list now but at least it won't work for clone table as you can see from the unit test.
2. This PR doesn't fix all the commands.

File: spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala [hunk 2]
```diff
@@ -274,8 +277,16 @@ class DeltaSqlParserSuite extends SparkFunSuite with SQLHelper {
...
+      tblName: String,
+      schema: String = null,
+      catalog: String = null): TableIdentifier = {
+    if (catalog == null) {
+      if (schema == null) new TableIdentifier(tblName)
+      else new TableIdentifier(tblName, Some(schema))
+    } else {
+      assert(schema != null)
+      new TableIdentifier(tblName, Some(schema), Some(catalog))
+    }
```
Comment (ryan-johnson-databricks) [inline spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala:290]: Isn't this just:
```suggestion
    new TableIdentifier(tblName, Option(schema), Option(catalog))
```
`TableIdentifier` constructor already verifies that schema is non-null when catalog is non-null:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala#L101

Comment (gengliangwang) [reply spark/src/test/scala/io/delta/sql/parser/DeltaSqlParserSuite.scala:290]: @ryan-johnson-databricks +1. We can update this later.

**Pull request closed without merging**